curvefs/test/client/client_s3_adaptor_Integration.cpp

```diff
@@ -0,0 +1,2849 @@
+        iter->second.buf = NULL;
+    }
+    gObjectDataMaps.clear();
+}
+
+TEST_F(ClientS3IntegrationTest, test_truncate_small3) {
+    curvefs::metaserver::Inode inode;
+
+    InitInodeForIntegration(&inode);
+    uint64_t offset = 0;
+    uint64_t len = 512 * 1024;
+    char *buf = new char[len];
+    memset(buf, 'a', len);
+    inode.set_length(8 * 1024 * 1024);
+    std::shared_ptr<FsCacheManager> fsCacheManager =
+        s3ClientAdaptor_->GetFsCacheManager();
+    uint64_t chunkId = 25;
+    EXPECT_CALL(mockMdsClient_, AllocS3ChunkId(_, _))
+        .WillRepeatedly(
+            DoAll(SetArgPointee<1>(chunkId), Return(FSStatusCode::OK)));
+    EXPECT_CALL(mockS3Client_, Upload(_, _, _))
+        .WillRepeatedly(Invoke(S3Upload));
+    EXPECT_CALL(mockS3Client_, Download(_, _, _, _))
+        .WillRepeatedly(Invoke(S3Download));
+    auto inodeWrapper = std::make_shared<InodeWrapper>(inode, nullptr);
+    EXPECT_CALL(mockInodeManager_, GetInode(_, _))
+        .WillRepeatedly(
+            DoAll(SetArgReferee<1>(inodeWrapper), Return(CURVEFS_ERROR::OK)));
+    EXPECT_CALL(mockS3Client_, UploadAsync(_))
+        .WillRepeatedly(
+            Invoke([&](const std::shared_ptr<PutObjectAsyncContext> &context) {
+                S3Data &tmp = gObjectDataMaps[context->key];
+                tmp.len = context->bufferSize;
+                tmp.buf = new char[context->bufferSize];
+                strncpy(tmp.buf, context->buffer, context->bufferSize);
+                context->retCode = 0;
+                context->cb(context);
+            }));
+
+
+    for (int i = 0; i < 6; i++) {
+        s3ClientAdaptor_->Write(inode.inodeid(), offset, len, buf);
+        offset = offset + len * 2;
+    }
+    ASSERT_EQ(6, fsCacheManager->GetDataCacheNum());
+    ASSERT_EQ(6 * len, fsCacheManager->GetDataCacheSize());
+    ASSERT_EQ(0, fsCacheManager->GetLruByte());
+
+    CURVEFS_ERROR ret = s3ClientAdaptor_->Flush(inode.inodeid());
+    inode = inodeWrapper->GetInodeUnlocked();
+    ASSERT_EQ(CURVEFS_ERROR::OK, ret);
+    ASSERT_EQ(0, fsCacheManager->GetDataCacheNum());
+    ASSERT_EQ(0, fsCacheManager->GetDataCacheSize());
+    ASSERT_EQ(6 * len, fsCacheManager->GetLruByte());
+
+    uint64_t truncateLen = 3 * 1024 * 1024;
+    ret = s3ClientAdaptor_->Truncate(&inode, truncateLen);
+    ASSERT_EQ(CURVEFS_ERROR::OK, ret);
+    ASSERT_EQ(3 * len, s3ClientAdaptor_->GetFsCacheManager()->GetLruByte());
+
+    truncateLen = 1200 * 1024;
+    ret = s3ClientAdaptor_->Truncate(&inode, truncateLen);
+    ASSERT_EQ(CURVEFS_ERROR::OK, ret);
+    ASSERT_EQ(len, s3ClientAdaptor_->GetFsCacheManager()->GetLruByte());
+
+    // cleanup
+    delete buf;
+    std::map<std::string, S3Data>::iterator iter = gObjectDataMaps.begin();
+    for (; iter != gObjectDataMaps.end(); iter++) {
+        delete iter->second.buf;
+        iter->second.buf = NULL;
+    }
+    gObjectDataMaps.clear();
+}
+
+TEST_F(ClientS3IntegrationTest, test_truncate_big1) {
```

Comment:
ditto. Please note the following similar test cases.

Reply:
fixed